IGNITE-14365 CDC Documentation (#9708)
apache/ignite: docs/_docs/persistence/change-data-capture.adoc

// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
= Change Data Capture

> WARNING: Experimental API.

Reviewer: ```suggestion
```

Author: Why you suggest to remove this warning?
CDC experimental  API, for now.

== Overview

> Change Data Capture (link:https://en.wikipedia.org/wiki/Change_data_capture[CDC]) is data processing pattern used to asynchronously receive entries that has changed on local node
> so that action can be taken using the changed entry.

Reviewer: ```suggestion
Change Data Capture (link:https://en.wikipedia.org/wiki/Change_data_capture[CDC]) is a data processing pattern used to asynchronously receive entries that have been changed on the local node so that action can be taken using the changed entry.
```

> 

Reviewer: ```suggestion

WARNING: Change Data Capture is an experimental feature whose API or design architecture might be changed.
```

> Some use-cases for CDC:

Reviewer: ```suggestion

Below are some of the Change Data Capture use cases:
```

>  * Stream changes in Warehouse.
>  * Update search index.
>  * Calculate statistics (streaming queries).
>  * Audit logs.
>  * Async interaction with extenal system - moderation, business process invocation, etc.

Reviewer: ```suggestion
 * Streaming changes in Warehouse;
 * Updating search index;
 * Calculating statistics (streaming queries);
 * Auditing logs;
 * Async interaction with extenal system: Moderation, business process invocation, etc.
```

> Ignite implements CDC with an application `ignite-cdc.sh` and link:https://github.com/apache/ignite/blob/master/modules/core/src/main/java/org/apache/ignite/cdc/CdcConsumer.java#L56[Java API].

Reviewer: ```suggestion
Ignite implements Change Data Capture with the `ignite-cdc.sh` application and link:https://github.com/apache/ignite/blob/master/modules/core/src/main/java/org/apache/ignite/cdc/CdcConsumer.java#L56[Java API].
```

> CDC application and Ignite node integrated via WAL archive segments:

Reviewer: ```suggestion
Below are the Change Data Capture application and the Ignite node integrated via WAL archive segments:
```

image:../../assets/images/integrations/CDC-design.svg[]

> When CDC enabled Ignite server node creates a hard link to each WAL archive segment in special `db/cdc/\{consistency_id\}` directory.
> `ignite-cdc.sh` run in different JVM and process newly archived link:native-persistence.adoc#_write-ahead_log[WAL segments].
> When segment fully processed by `ignite-cdc.sh` it removed. The actual disk space free when both links - archive and CDC removed.
> tate of consumption can be saved to disk to continue from it in the case of any failure.

Reviewer: ```suggestion
When Change Data Capture is enabled, the Ignite server node creates a hard link to each WAL archive segment in the special `db/cdc/\{consistency_id\}` directory.
The `ignite-cdc.sh` application runs on a different JVM and processes newly archived link:native-persistence.adoc#_write-ahead_log[WAL segments].
When the segment is fully processed by `ignite-cdc.sh`, it is removed. The actual disk space is free when both links (archive and Change Data Capture) are removed.
The state of consumption can be saved to continue from it in case of any failure.
```

> tate of consumption can be saved to disk to continue from it in the case of any failure.

Reviewer: Personally, I would rework this sentence as it's not clear what the "State of consumption" concept is.

Author: > State of consumption is a pointer to the last processed event.
> Consumer can tell to `ignite-cdc.sh` to save the consumption state.
> On startup event processing will be continued from the last saved state.

Is it better?

== Configuration

> === Ignite node:

Reviewer: ```suggestion
=== Ignite Node
```

[cols="20%,45%,35%",opts="header"]
|===
|Name |Description | Default value
| `DataStorageConfiguration#cdcEnabled` | Flag to enable CDC on server node. | `false`
| `DataStorageConfiguration#cdcWalPath` | Path to the CDC directory | `"db/wal/cdc"`
| `DataStorageConfiguration#walForceArchiveTimeout` | Timeout to forcefully archive WAL segment even it not full. | `-1` (disabled)
|===

> === CDC application:

Reviewer: ```suggestion
=== Change Data Capture Application
```

> CDC configured in the same manner as Ignite node via spring XML file.
> `ignite-cdc.sh` require both Ignite and CDC configuration to start.
> `IgniteConfiguration` used to determine common options like path to CDC directory, node consistent id and others.
> And `CdcConfiguration` contains options specifically for `ignite-cdc.sh`.

Reviewer: ```suggestion
Change Data Capture is configured in the same way as the Ignite node - via the spring XML file:

* `ignite-cdc.sh` requires both Ignite and Change Data Capture configurations to start;
* `IgniteConfiguration` is used to determine common options like a path to the Change Data Capture directory, node consistent id, and other parameters;
* `CdcConfiguration` contains `ignite-cdc.sh`-specific options.
```

[cols="20%,45%,35%",opts="header"]
|===

> |Name |Description | Default value
> | `lockTimeout` | Timeout to wait lock acquire. CDC locks directory on startup to ensure there is no concurrent CDC processing same directory.
> | 1000 milliseconds.
> | `checkFrequency` | Amount of time application sleeps between subsequent checks when no new files available. | 1000 milliseconds.
> | `keepBinary` | Flag to specify if key and value of changed entries should be provided in link:../key-value-api/binary-objects.adoc[binary format]. | `true`
> | `consumer` | Implementation of `org.apache.ignite.cdc.CdcConsumer` that consumes entries changes | null
> | `metricExporterSpi` | Array of SPI's to export CDC metrics. See link:../monitoring-metrics/new-metrics-system.adoc#_metric_exporters[metrics] documentation, also. | null

Reviewer: ```suggestion
|Name |Description | Default value
| `lockTimeout` | Timeout to wait for lock acquiring. Change Data Capture locks directory on a startup to ensure there is no concurrent Change Data Capture processing in the same directory.
| 1000 milliseconds.
| `checkFrequency` | Amount of time application sleeps between subsequent checks when no new files available. | 1000 milliseconds.
| `keepBinary` | Flag to specify if key and value of changed entries should be provided in link:../key-value-api/binary-objects.adoc[binary format]. | `true`
| `consumer` | Implementation of `org.apache.ignite.cdc.CdcConsumer` that consumes entries changes | null
| `metricExporterSpi` | Array of SPI's to export CDC metrics. See link:../monitoring-metrics/new-metrics-system.adoc#_metric_exporters[metrics] documentation, also. | null
```

* `ignite-cdc.sh` requires both Ignite and Change Data Capture configurations to start;
* `IgniteConfiguration` is used to determine common options like a path to the Change Data Capture directory, node consistent id, and other parameters;
* `CdcConfiguration` contains `ignite-cdc.sh`-specific options.

[cols="20%,45%,35%",opts="header"]
|===
|Name |Description | Default value
| `lockTimeout` | Timeout to wait for lock acquiring. Change Data Capture locks directory on a startup to ensure there is no concurrent Change Data Capture processing in the same directory.
| 1000 milliseconds.
| `checkFrequency` | Amount of time application sleeps between subsequent checks when no new files available. | 1000 milliseconds.
| `keepBinary` | Flag to specify if key and value of changed entries should be provided in link:../key-value-api/binary-objects.adoc[binary format]. | `true`
| `consumer` | Implementation of `org.apache.ignite.cdc.CdcConsumer` that consumes entries changes | null
| `metricExporterSpi` | Array of SPI's to export CDC metrics. See link:../monitoring-metrics/new-metrics-system.adoc#_metric_exporters[metrics] documentation, also. | null
|===

== API

> === `org.apache.ignite.cdc.CdcEvent`
> Single change of data reflected by `CdcEvent`.
> Let's get a closer look to the information provided:
> 
> [cols="20%,80%",opts="header"]
> |===
> |Name |Description
> | `key()` | Key for the changed entry.
> | `value()` | Value for the changed entry. This method will return `null` if event reflects removal.
> | `cacheId()` | ID of the cache where change happens. Value equal to the `CACHE_ID` from `SYS.CACHES` link:../monitoring-metrics/system-views.adoc#_CACHES[system view]
> | `partition()` | Partition of the changed entry.
> | `primary()` | Flag to distinguish if operation happen on primary or backup node.
> | `version()` | `Comparable` version of the changed entry. Internally Ignite maintain ordered versions of each entry so any changes of the same entry can be sorted.
> |===

Reviewer: ```suggestion
=== `org.apache.ignite.cdc.CdcEvent`

Below is a single change of the data reflected by `CdcEvent`. 

[cols="20%,80%",opts="header"]
|===
|Name |Description
| `key()` | Key for the changed entry.
| `value()` | Value for the changed entry. This method will return `null` if the event reflects removal.
| `cacheId()` | ID of the cache where the change happens. The value is equal to the `CACHE_ID` from `SYS.CACHES` link:../monitoring-metrics/system-views.adoc#_CACHES[system view]
| `partition()` | Partition of the changed entry.
| `primary()` | Flag to distinguish if operation happens on the primary or a backup node.
| `version()` | `Comparable` version of the changed entry. Internally, Ignite maintains ordered versions of each entry so any changes of the same entry can be sorted.
|===
```

=== `org.apache.ignite.cdc.CdcConsumer`

Consumer of change events. Should be implemented by the user.
[cols="20%,80%",opts="header"]
|===
|Name |Description
| `void start(MetricRegistry)` | Invoked one time on the start of CDC application. `MetricRegistry` should be used to export consumer specific metrics.
| `boolean onEvents(Iterator<CdcEvent> events)` | Main method that processes changes. When this method returns `true` state will be saved on the disk. State points to the event next to the last read. In case of any failure consumption will continue from the last saved state.
| `void stop()` | Invoked one time on the stop of CDC application.
|===

== Metrics

> `ignite-cdc.sh` uses the same SPI to export metrics as Ignite.

Reviewer: ```suggestion
`ignite-cdc.sh` uses the same SPI to export metrics as Ignite does.
```

> The following metrics provided by the application itself (additional metrics may be provided by the consumer):

Reviewer: ```suggestion
The following metrics are provided by the application (additional metrics can be provided by the consumer):
```

|===
|Name |Description

> | CurrentSegmentIndex | Index of currently processed WAL segment.
> | CommittedSegmentIndex | Index of WAL segment contains last commited state.
> | CommittedSegmentOffset | Commited offset in bytes inside WAL segment.
> | LastSegmentConsumptionTime | Time in milliseconds when last segment processing was started.
> | BinaryMetaDir | Binary meta directory this application reads from.
> | MarshallerDir | Marshaller directory this application reads from.
> | CdcDir | CDC directory this application reads from.

Reviewer: ```suggestion
| CurrentSegmentIndex | Index of the currently processing WAL segment.
| CommittedSegmentIndex | Index of the WAL segment that contains the last committed state.
| CommittedSegmentOffset | Committed offset in bytes inside the WAL segment.
| LastSegmentConsumptionTime | Timestamp (in milliseconds) indicating the last segment processing start.
| BinaryMetaDir | Binary meta-directory the application reads data from.
| MarshallerDir | Marshaller directory the application reads data from.
| CdcDir | The Change Data Capture directory the application reads data from.
```

|===

== Logging

> `ignite-cdc.sh` use the same logging configuration as Ignite node. The only difference is log written in file "ignite-cdc.log"

Reviewer: ```suggestion
`ignite-cdc.sh` uses the same logging configuration as the Ignite node does. The only difference is that the log is written in the"ignite-cdc.log" file.
```

== Lifecycle

> IMPORTANT: `ignite-cdc.sh` implements fail-fast approach. It will just fail in case of any error. Restart should be configured with the OS tools.

Reviewer: ```suggestion
IMPORTANT: `ignite-cdc.sh` implements the fail-fast approach. It just fails in case of any error. The restart procedure should be configured with the OS tools.
```

>  1. Finds required shared directories. Values from provided `IgniteConfiguration` taken.
>  2. Locks CDC directory.
>  3. Loads saved state.
>  4. Starts the consumer.
>  5. Infinitely waits for new available segment and processes it.
>  6. Stop the consumer in case of failure or receive stop signal.

Reviewer: ```suggestion
 1. Find the required shared directories. Take the values from the provided `IgniteConfiguration`.
 2. Lock the Change Data Capture directory.
 3. Load the saved state.
 4. Start the consumer.
 5. Infinitely wait for the newly available segment and process it.
 6. Stop the consumer in case of a failure or a received stop signal.
```
